sync/src/main/java/tech/pegasys/artemis/sync/SyncManager.java

```diff
@@ -13,60 +13,102 @@
 
 package tech.pegasys.artemis.sync;
 
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+import com.google.common.annotations.VisibleForTesting;
 import com.google.common.primitives.UnsignedLong;
 import java.util.Comparator;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import tech.pegasys.artemis.networking.eth2.Eth2Network;
 import tech.pegasys.artemis.networking.eth2.peers.Eth2Peer;
-import tech.pegasys.artemis.statetransition.BlockImporter;
+import tech.pegasys.artemis.service.serviceutils.Service;
 import tech.pegasys.artemis.storage.ChainStorageClient;
 
-public class SyncManager {
-
+public class SyncManager extends Service {
+  private static final Logger LOG = LogManager.getLogger();
   private final Eth2Network network;
   private final ChainStorageClient storageClient;
-  private final BlockImporter blockImporter;
+  private final PeerSync peerSync;
 
-  private final AtomicBoolean syncActive = new AtomicBoolean(false);
+  private boolean syncActive = false;
+  private boolean syncQueued = false;
+  private volatile long peerConnectSubscriptionId;
 
   public SyncManager(
-      final Eth2Network network,
-      final ChainStorageClient storageClient,
-      final BlockImporter blockImporter) {
+      final Eth2Network network, final ChainStorageClient storageClient, final PeerSync peerSync) {
     this.network = network;
     this.storageClient = storageClient;
-    this.blockImporter = blockImporter;
-    network.subscribeConnect(this::onNewPeer);
+    this.peerSync = peerSync;
   }
 
-  public CompletableFuture<Void> sync() {
-    if (!syncActive.compareAndSet(false, true)) {
-      return CompletableFuture.failedFuture(new RuntimeException("Sync already active"));
-    }
-    return executeSync().thenRun(() -> syncActive.set(false));
+  @Override
+  protected CompletableFuture<?> doStart() {
+    peerConnectSubscriptionId = network.subscribeConnect(this::onNewPeer);
+    startOrScheduleSync();
+    return completedFuture(null);
   }
 
-  private CompletableFuture<Void> executeSync() {
-    Optional<Eth2Peer> possibleSyncPeer = findBestSyncPeer();
+  @Override
+  protected CompletableFuture<?> doStop() {
+    network.unsubscribeConnect(peerConnectSubscriptionId);
+    synchronized (this) {
+      syncQueued = false;
+    }
+    return completedFuture(null);
```

Comment:
Should we make sure our last request completes or is cancelled?  I think it's possible that we could keep sending out requests if there's an outstanding request that completes unsuccessfully and causes us to retry, potentially multiple times.

Reply:
Good spot. Done.